src/webgpu/shader/execution/memory_model/memory_model_setup.ts

```diff
@@ -0,0 +1,511 @@
+  stressLineSize: number;
+  /** The number of blocks of memory to assign stressing threads to. */
+  stressTargetLines: number;
+  /** How non-testing threads are assigned to stressing locations. 100 means all iterations use a round robin approach, 0 means all use a chunking approach. */
+  stressStrategyBalancePct: number;
+  /** Used to permute thread ids within a workgroup, so more random pairings are created between threads coordinating on a test. */
+  permuteFirst: number;
+  /** Used to create distance between memory locations used in a test. Set this to 1 for memory that should be aliased. */
+  permuteSecond: number;
+  /** The distance (in number of 4 byte intervals) between any two memory locations used for testing. */
+  memStride: number;
+  /** For tests that access one memory location, but use dynamic addresses to avoid compiler optimization, aliased memory should be set to true. */
+  aliasedMemory: boolean;
+  /** The number of memory locations accessed by this test. */
+  numMemLocations: number;
+  /** The number of read outputs per test that need to be analyzed in the result aggregation shader. */
+  numReadOutputs: number;
+  /** The number of possible behaviors that a test can have. */
+  numBehaviors: number;
+};
+
+/** Represents a device buffer and a utility buffer for resetting memory and copying parameters. */
+type BufferWithSource = {
+  deviceBuf: GPUBuffer;
+  srcBuf: GPUBuffer;
+  size: number;
+};
+
+/** Specifies the buffers used during a memory model test. */
+type MemoryModelBuffers = {
+  /** This is the memory region that testing threads read from and write to. */
+  testLocations: BufferWithSource;
+  /** This buffer collects the results of reads for analysis in the result aggregation shader. */
+  readResults: BufferWithSource;
+  /** This buffer is the aggregated results of every testing thread, and is used to check for test success/failure. */
+  testResults: BufferWithSource;
+  /** This buffer stores the shuffled workgroup ids for use during testing. */
+  shuffledWorkgroups: BufferWithSource;
+  /** This is the bounded spin-loop barrier, used to temporally align testing threads. */
+  barrier: BufferWithSource;
+  /** Memory region for stressing threads to read to and write from. */
+  scratchpad: BufferWithSource;
+  /** The memory locations in the scratch region that stressing threads access. */
+  scratchMemoryLocations: BufferWithSource;
+  /** Parameters that are used by the shader to calculate memory locations and perform stress. */
+  stressParams: BufferWithSource;
+};
+
+/** The number of stress params to add to the stress params buffer. */
+const numStressParams = 12;
+
+/** Buffer sizes are specified in number of memory locations, multiplying by 4 gives the number of bytes. */
+const byteMultiplier = 4;
+
+/** Uniform buffer elements must align to a 16 byte boundary (see https://www.w3.org/TR/WGSL/#storage-class-layout-constraints). */
```

Comment:
This is a misreading of the spec. Only arrays and structs have a requirement to be aligned to a 16 byte multiple. You could change all the elements in your parameter structure in the shader to drop `@size(16)`.

Reply:
ah yeah, originally I had all the parameters in an array, and I guess I forgot to look into whether that was needed. Removed the need for 16 byte alignment